### PR TITLE
Put specific nibabel code in a separate module

### PR DESCRIPTION
**Pull request by AlexandreAbraham**
This PR is a first step toward fixing #380. I have simply moved all nibabel related code (except for test code and a special hack in MultiPCA) to an indenpendant module. Since this code is already tested by most of nilearn functions, I simply added a test for a special case that wasn't covered.

I think that the discussion will mostly focus on the naming conventions. I'd like to merge it fast as it will massively impact further PR. We should also be careful for the upcoming PR: nibabel must never be imported except for test (because nilearn input are nibabel objects) and specific Nifti1Image manipulation (in the latter case, the code should be in the dedicated module).

### CODE REVIEW

**Comment by bcipolli** [comment]
I would suggest to name it `_new_img`, if it's an internal-only / hidden API.  Or, if it's planned to recommend users to use this function, I suggest adding a docstring to `new_img` about what object type is returned.

**Comment by AlexandreAbraham** [comment]
It is already located in `utils` which is hidden. However, I must admit that I didn't get the naming conventions in there (for example `_get_shape`).

File: nilearn/_utils/niimage.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    # We use a try/except here because this is the way hasattr works
+    try:
+        get_data = getattr(obj, "get_data")
+        get_affine = getattr(obj, "get_affine")
+        return callable(get_data) and callable(get_affine)
+    except AttributeError:
+        return False
+
+
+def load_img(img):
```
Comment (bthirion) [inline nilearn/_utils/niimage.py]: dosctring ?

File: nilearn/_utils/niimage.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            return "%s(\nshape=%s,\naffine=%s\n)" % \
+                   (niimgs.__class__.__name__,
+                    repr(_get_shape(niimgs)),
+                    repr(niimgs.get_affine()))
+    except:
+        pass
+    return repr(niimgs)
+
+
+def short_repr(niimg):
```
Comment (bthirion) [inline nilearn/_utils/niimage.py]: dosctring ?

**Comment by bthirion** [comment]
Looks good otherwise. I don't see any travis output ?

**Comment by bthirion** [comment]
There are still many calls to Nifti1Image throughout all the codebase...

**Comment by AlexandreAbraham** [comment]
Yes, I noticed that on my way back home ;). I fixed the last occurences.

**Comment by lesteve** [comment]
Naming suggestions:

- maybe something like create_img or make_img ?

- about the filename why not nibabel_image.py rather than niimage.py to make it more explicit that it is a nibabel wrapper? Also niimage.py doesn't follow the same naming convention than niimg_conversions.py in _utils

**Comment by AlexandreAbraham** [comment]
> maybe something like create_img or make_img ?

Not a big fan, I prefer `new_img`.

> why not nibabel_image.py

Because we don't want to be tied to nibabel. If one day we change the image format nilearn relies on, all we will have to do is to change the dependency on that file.

> Also niimage.py doesn't follow the same naming convention than niimg_conversions.py in _utils

I agree, `niimg_conversions` is definitely a bad name ;)

**Comment by AlexandreAbraham** [comment]
I am going to rename niimage -> niimg (but it could be img)

@GaelVaroquaux any thoughts on this one? I'd like to merge it fast as it is just code refactoring.

File: nilearn/_utils/niimage.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
+import copy
```
Comment (GaelVaroquaux) [inline nilearn/_utils/niimage.py]: Can you give us a docstring that says what's in this module?

Also license information as a comment.

File: nilearn/_utils/niimage.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        get_data = getattr(obj, "get_data")
+        get_affine = getattr(obj, "get_affine")
+        return callable(get_data) and callable(get_affine)
+    except AttributeError:
+        return False
+
+
+def load_img(img):
+    """Load a niimg and check if it has required methods
+    """
```
Comment (GaelVaroquaux) [inline nilearn/_utils/niimage.py]: Please define input and output types in the docstring.

File: nilearn/_utils/niimage.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        img = nibabel.load(img)
+    elif not is_img(img):
+        raise TypeError("Data given cannot be converted to a nifti"
+                        " image: this object -'%s'- does not expose"
+                        " get_data or get_affine methods"
+                        % short_repr(img))
+    return img
+
+
+def new_img(data, affine, header=None):
```
Comment (GaelVaroquaux) [inline nilearn/_utils/niimage.py]: I tend to have a strong bias against such a wrapper function that does nothing more than the function it wraps. The reason is that it makes the code opaque.

File: nilearn/_utils/niimage.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        copy of input (data and affine)
+    """
+    if not is_img(img):
+        raise ValueError("Input value is not an image")
+    return nibabel.Nifti1Image(img.get_data().copy(),
+                               img.get_affine().copy())
+
+
+def save_img(img, filename):
+    img.to_filename(filename)
```
Comment (GaelVaroquaux) [inline nilearn/_utils/niimage.py]: Why? Why such a  one-liner. Why not teach our users to use 'to_filename' (which is a pretty bad method name, but let's move forward.

File: nilearn/_utils/niimage.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
+import copy
+import gc
+import collections
+
+import nibabel
+
+
+def _get_shape(img):
```
Comment (banilo) [inline nilearn/_utils/niimage.py]: General remark: there appears to be quite some inconsistency in when and how this function is used in nilearn code.

File: nilearn/_utils/niimage.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+def _safe_get_data(img):
+    """ Get the data in the image without having a side effect on the
+        Nifti1Image object
+    """
+    if hasattr(img, '_data_cache') and img._data_cache is None:
+        # Copy locally the Nift1Image object to avoid the side effect of data
+        # loading
+        img = copy.deepcopy(img)
+    # typically the line below can double memory usage
```
Comment (banilo) [inline nilearn/_utils/niimage.py]: rather "line above"?

File: nilearn/_utils/niimage.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+    # that's why we invoke a forced call to the garbage collector
+    gc.collect()
+    return img.get_data()
+
+
+def is_img(obj):
```
Comment (banilo) [inline nilearn/_utils/niimage.py]: I might have asked this question before: Why is there no is_niimg(obj) in nilearn? ...one that verifies whether the object is Nifti1Image/iterable/list?

File: nilearn/_utils/niimage.py (hunk 6 above)
Comment (banilo) [inline nilearn/_utils/niimage.py]: > Why such a one-liner.

+1

File: nilearn/plotting/img_plotting.py [hunk 10]
```diff
@@ -76,7 +75,7 @@ def _plot_img_with_bg(img, bg_img=None, cut_coords=None,
             # voxels pass the threshold
             threshold = fast_abs_percentile(data) - 1e-5
 
-        img = nibabel.Nifti1Image(as_ndarray(data), affine)
+        img = new_img(as_ndarray(data), affine)
```
Comment (banilo) [inline nilearn/plotting/img_plotting.py]: I like new_img() already.

File: nilearn/_utils/niimage.py (hunk 7 above)
Comment (AlexandreAbraham) [inline nilearn/_utils/niimage.py]: You mean that it is sometimes used from `niimg_conversion` module? I've seen that in the tests, I will fix it.

File: nilearn/_utils/niimage.py (hunk 8 above)
Comment (AlexandreAbraham) [inline nilearn/_utils/niimage.py]: No, I think that this comment refer to the last line of the function.

File: nilearn/_utils/niimage.py (hunk 9 above)
Comment (AlexandreAbraham) [inline nilearn/_utils/niimage.py]: I think it's because it boils down to loading the data itself so doing the equivalent of `check_niimg`.

File: nilearn/_utils/niimage.py (hunk 5 above)
Comment (AlexandreAbraham) [inline nilearn/_utils/niimage.py]: As said in the description, this PR does nothing but moving the code around. Since nibabel related functions were scattered all around the code, I moved them all there. While doing that, I seized the occasion to remove all `nibabel` related code and put aliases here. I think this is a good practice to _abstract_ this dependency but I can revert it back to what it was.

File: nilearn/_utils/niimage.py (hunk 6 above)
Comment (AlexandreAbraham) [inline nilearn/_utils/niimage.py]: This is the purpose of the PR: remove any reference to nibabel all around the code. A `niimg` being an object exposing `get_data()` and `get_affine()` method, I wrapped the call to `to_filename()`.

Comment (GaelVaroquaux) [inline nilearn/_utils/niimage.py]: > This is the purpose of the PR: remove any reference to nibabel all around the
> code.

I don't think that this is a worthwhile goal per-se: it will simply make
the code opaque. It is good thing to explicitely know that we are using
nibabel.

Comment (AlexandreAbraham) [inline nilearn/_utils/niimage.py]: OK, so I can close this PR and we let the code related to nifti images in niimg_conversion.

Comment (GaelVaroquaux) [inline nilearn/_utils/niimage.py]: > OK, so I can close this PR and we let the code related to nifti images in
> niimg_conversion.

Well, maybe this PR was bringing in useful cleanups?

Comment (AlexandreAbraham) [inline nilearn/_utils/niimage.py]: There is no code engineering in this PR, only moving around some code (mainly from niimg_conversion to niimage) and adding a test. If we don't want to isolate nibabel in a file, there is no reason to refactor the code. If we want to keep something from this PR, I think it would be easier to create a new (clean) one.

Comment (GaelVaroquaux) [inline nilearn/_utils/niimage.py]: I have no issue with moving code around. It's just the thin abstractions
introduced that I don't like. Based on my experience of discovering a new
codebase, they render a codebase opaque to an outsider.

File: nilearn/_utils/niimage.py (hunk 5 above)
Comment (eickenberg) [inline nilearn/_utils/niimage.py]: Once the header-related `new_img_like` (or however it will be called) becomes reality, the function name `new_img` will probably be looked for by users that know the former. Yes, this is `__future__` talk :) but in that light this addition seems very much consistent.

Comment (GaelVaroquaux) [inline nilearn/_utils/niimage.py]: > Once the header-related new_img_like (or however it will be called) becomes
> reality, the function name new_img will probably be looked for by users that
> know the former. Yes, this is __future__ talk :) but in that light this
> addition seems very much consistent.

OK, then let's tackle this problem now. I don't want to design abstractions
for future usage, as this is a very slippery slope.

Comment (AlexandreAbraham) [inline nilearn/_utils/niimage.py]: > Once the header-related new_img_like (or however it will be called) becomes reality, the function name new_img will probably be looked for by users that know the former.

Do you plan to make `new_img_like` available to the users? I thought that it would be internal cuisine.

Comment (GaelVaroquaux) [inline nilearn/_utils/niimage.py]: > Do you plan to make new_img_like available to the users? I thought that it
> would be internal cuisine.

Agreed.

**Comment by eickenberg** [comment]
In reading through all the changed examples, I feel that `new_img` is actually really nice on the eyes. 

@GaelVaroquaux I totally understand your apprehension about thin wrappers since they definitely add a layer, however thin it is. As somebody who has difficulties getting into new codebases and their practices, this is definitely an added one.

But really, `new_img` seems very elegant to me, at least much more elegant than `nibabel.Nifti1Image` -- which to new users may actually be equally shocking simply by its unwieldiness.

`nibabel.Nifti1Image` still works for anybody who wants to use it, and one can always see `new_img` as a simple alias, so I really think this change is worth it.

Wrt `new_img_like`, maybe tackling it right now is a good idea. Do we need a discussion about the `mode=` that had been put forward? (I am not sure which ones there would be.)

**Comment by AlexandreAbraham** [comment]
You would achieve the same effect by doing: `from nibabel import Nifti1Image as niimg` for example. Aliasing is not the best practice but it is maybe better than a wrapper.

**Comment by eickenberg** [comment]
I realized that while typing the word *alias*, originally intended to be more figurative than real. Small caps for a class is not cool, and `Niimg` ... would look weird wrt all the other true functions that contain the word `niimg`. I think I am overthinking this.

**Comment by AlexandreAbraham** [comment]
OK, this PR is back on track again (because I like to nilearn a bit after lunch). I'm afraid that my renaming of the file erased the comments but I will address them when I have time.

Changes:
* Remove the one-liner `save_img` (since SpatialImage provides the `to_filename` method
* Refactor the one-liner `new_img` into `new_img_like` that actually does something. The constructor of the original type of the image is used to be able to support any type (this is the way nibabel handles it). I also put in there some small tweaks.
* Propagate the changes through all the code. Tests are passing but review is welcome as I sometimes had to change things a little bit.

**Comment by AlexandreAbraham** [comment]
Rebased on master for python3 compliance.

File: nilearn/_utils/niimg.py [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+from six import string_types
+
+import numpy as np
+
+import nibabel
+
+from .numpy_conversions import as_ndarray
+
+
+def _get_shape(img):
```
Comment (AlexandreAbraham) [inline nilearn/_utils/niimg.py]: Since SpatialImage should provide `shape` property, I am in favor of removing this one.

File: nilearn/_utils/niimg.py [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+        # Copy locally the Nift1Image object to avoid the side effect of data
+        # loading
+        img = copy.deepcopy(img)
+    # typically the line below can double memory usage
+    # that's why we invoke a forced call to the garbage collector
+    gc.collect()
+    return img.get_data()
+
+
+def is_img(obj):
```
Comment (AlexandreAbraham) [inline nilearn/_utils/niimg.py]: This should be replaced by `isinstance(nibabel.spatialimages.SpatialImage).

File: nilearn/_utils/niimg.py [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+        Transformation matrix
+
+    copy_header: boolean, optional
+        Indicated if the header of the reference image should be used to
+        create the new image
+
+    Returns
+    -------
+
+    new_img: nibabel.SpatialImage
```
Comment (AlexandreAbraham) [inline nilearn/_utils/niimg.py]: I am not satisfied with that but I don't know how to say here that the type is depending on an input.

File: nilearn/_utils/niimg.py [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+        nibabel.Nifti1Image object to copy.
+
+    Returns
+    =======
+    img_copy: nibabel.Nifti1Image
+        copy of input (data and affine)
+    """
+    if not is_img(img):
+        raise ValueError("Input value is not an image")
+    return nibabel.Nifti1Image(img.get_data().copy(),
```
Comment (AlexandreAbraham) [inline nilearn/_utils/niimg.py]: This is wrong.

File: nilearn/_utils/testing.py [hunk 15]
```diff
@@ -424,44 +423,46 @@ def generate_fake_fmri(shape=(10, 11, 12), length=17, kind="noise",
...
-        flat_fmri /= np.abs(flat_fmri).max()
-        target = np.zeros(length, dtype=np.int)
-        rest_max_size = (length - (n_blocks * block_size)) // n_blocks
-        if rest_max_size < 0:
-            raise ValueError(
-                '%s is too small '
-                'to put %s blocks of size %s' % (
-                    length, n_blocks, block_size))
-        t_start = 0
+    if n_blocks is None:
```
Comment (AlexandreAbraham) [inline nilearn/_utils/testing.py]: Just cosmit refactoring: I return in this condition instead of having a huge condition block.

File: nilearn/image/image.py [hunk 16]
```diff
@@ -429,44 +427,36 @@ def mean_img(imgs, target_affine=None, target_shape=None,
...
-        total_n_imgs = 1
-    else:
-        try:
-            total_n_imgs = len(imgs)
-        except:
-            total_n_imgs = None
 
     imgs_iter = iter(imgs)
-
+    first_img = next(imgs_iter)
```
Comment (AlexandreAbraham) [inline nilearn/image/image.py]: Treating the first file first each time greatly simplify the code.

File: nilearn/image/resampling.py [hunk 17]
```diff
@@ -185,7 +186,7 @@ def get_mask_bounds(img):
 
     """
     img = _utils.check_niimg(img)
-    mask = img.get_data()
+    mask = (img.get_data() != 0.)
```
Comment (AlexandreAbraham) [inline nilearn/image/resampling.py]: We have to ensure that the mask is bool.

Comment (GaelVaroquaux) [inline nilearn/image/resampling.py]: > We have to ensure that the mask is bool.

I think that sklearn.utils.fixes.as_type(mask, np.bool) would be more
memory-friendly.

Comment (AlexandreAbraham) [inline nilearn/image/resampling.py]: Memory consumption is the same (we need a boolean copy of img.get_data()). From my experience, direct operations are faster than `astype`.

Comment (GaelVaroquaux) [inline nilearn/image/resampling.py]: > Memory consumption is the same (we need a boolean copy of
> img.get_data()).

Because it is impossible to have booleans in nifti, right? Elsewhere it
would indeed be different.

That just reminded me what is the "right" way to do it in the nilearn
codebase: _utils.numpy_conversions._asarray(arr, dtype=np.bool). This
code does some tricks to avoid a memory copy when converting from int8 to
bool. Philippe wrote that because he was annoyed that we couldn't store
booleans in nifti files.

Comment (AlexandreAbraham) [inline nilearn/image/resampling.py]: Fixed.

**Comment by AlexandreAbraham** [comment]
Rebased. Any other objection on this one? I think that I've addressed all the comments and that it is now actually a step forward cleaning code in nilearn ;).

File: nilearn/_utils/niimg.py (hunk 11 above)
Comment (eickenberg) [inline nilearn/_utils/niimg.py]: Is this still necessary?

File: nilearn/_utils/niimg.py (hunk 13 above)
Comment (eickenberg) [inline nilearn/_utils/niimg.py]: How about: `new instance of the input image class, whose API is compatible nibabel.SpatialImage. See docs: `

Comment (eickenberg) [inline nilearn/_utils/niimg.py]: maybe `type` is better, as you describe. My main point is that it doesn't say `new_img: nibabel.SpatialImage`, at least put the word 'compatible' in there somewhere.

Comment (eickenberg) [inline nilearn/_utils/niimg.py]: forget what I said. Deleting comments.

File: nilearn/_utils/niimg.py [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+        filename = niimgs.get_filename()
+        if filename is not None:
+            return "%s('%s')" % (niimgs.__class__.__name__,
+                                filename)
+        else:
+            return "%s(\nshape=%s,\naffine=%s\n)" % \
+                   (niimgs.__class__.__name__,
+                    repr(_get_shape(niimgs)),
+                    repr(niimgs.get_affine()))
+    except:
```
Comment (lesteve) [inline nilearn/_utils/niimg.py]: Should the exception type be a bit more specific, like AttributeError?

File: nilearn/plotting/img_plotting.py [hunk 19]
```diff
@@ -208,6 +208,11 @@ class _MNI152Template(object):
     vmax   = None
     _shape  = None
 
+    def __init__(self, *args, **kwargs):
+        # This looks like a hack but we want a constant image compatible with
+        # all interfaces
+        pass
```
Comment (lesteve) [inline nilearn/plotting/img_plotting.py]: Sorry I don't get this, the only use I could find of using _MNI152Template was without any arguments in the `__init__`.

Also just curious, what's the advantage of having _MNI152Template be an image?

File: nilearn/_utils/niimg.py [hunk 20]
```diff
@@ -0,0 +1,168 @@
...
+    if hasattr(img, 'shape'):
+        shape = img.shape
+    else:
+        shape = img.get_data().shape
+    return shape
+
+
+def _safe_get_data(img):
+    """ Get the data in the image without having a side effect on the
+        Nifti1Image object
```
Comment (lesteve) [inline nilearn/_utils/niimg.py]: Should "Nifti1Image object" be replaced by "Image object" or "SpatialImage object"?

**Comment by lesteve** [comment]
By the way the error you are seeing in Travis has been fixed in master so you probably want to do a rebase.

File: nilearn/_utils/niimg.py (hunk 12 above)
Comment (lesteve) [inline nilearn/_utils/niimg.py]: I agree unless someone comes up with a compelling use case of using an `img` which is not a `nibabel.SpatialImage`.

File: nilearn/_utils/niimg.py [hunk 21]
```diff
@@ -0,0 +1,168 @@
...
+                        % short_repr(img))
+    return img
+
+
+def new_img_like(ref_img, data, affine, copy_header=False):
+    """Create a new image of the same class as the reference image
+
+    Parameters
+    ----------
+    ref_img: Niimg-like image
```
Comment (lesteve) [inline nilearn/_utils/niimg.py]: Can you use "Niimg-like object" which is the convention we tried to use everywhere else previously? Bonus points if you add the doc URL too.

File: nilearn/_utils/niimg.py [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+    -------
+
+    new_img: nibabel.SpatialImage
+        An image which has the same type as the reference image.
+    """
+    # XXX If the niimg is a list of 3D images, we don't need to load them all
+    # But this is a bit ugly
+    if hasattr(ref_img, "__iter__") and not isinstance(ref_img, string_types):
+        ref_img = next(ref_img)
+    ref_img = load_img(ref_img)
```
Comment (lesteve) [inline nilearn/_utils/niimg.py]: I assumed that new_img_like would only take an `img` object rather than a niimg-like one. Is this added fucntionality actually used in the code?

File: nilearn/_utils/niimg.py [hunk 23]
```diff
@@ -0,0 +1,157 @@
...
+from six import string_types
+
+import numpy as np
+import nibabel
+
+from .numpy_conversions import as_ndarray
+
+
+def _safe_get_data(img):
+    """ Get the data in the image without having a side effect on the
```
Comment (AlexandreAbraham) [inline nilearn/_utils/niimg.py]: Does anybody know what problem solves this function? From what I understand, if the data is memmaped, it loads it into memory. First:
* Why is this behavior needed?
* Why not using directly the functions provided by nibabel to do so?

File: nilearn/_utils/niimg.py (hunk 20 above)
Comment (AlexandreAbraham) [inline nilearn/_utils/niimg.py]: Not sure if all SpatialImage have a cache...

File: nilearn/_utils/niimg.py (hunk 12 above)
Comment (AlexandreAbraham) [inline nilearn/_utils/niimg.py]: Ok, I'll wait a bit before changing that part.

File: nilearn/_utils/niimg.py (hunk 23 above)
Comment (GaelVaroquaux) [inline nilearn/_utils/niimg.py]: > Does anybody know what problem solves this function? From what I understand, if
> the data is memmaped, it loads it into memory. First:

I think that the goal is to prevent memory leaks: when 'get_data' is
called, the niftiimage object has the data attached to it. If a reference
is kept on this object, outside of the function, we have a memory leak,
as the function just loaded data that won't be removed.

File: nilearn/_utils/niimg.py (hunk 22 above)
Comment (AlexandreAbraham) [inline nilearn/_utils/niimg.py]: No idea what is the difference between an `img` and a `niimg-like`.

Comment (lesteve) [inline nilearn/_utils/niimg.py]: > No idea what is the difference between an img and a niimg-like.

Hint: one is almost like the other. More seriously `img` is passing the `is_img` check (so it is an object with get_data and get_affine at the moment) whereas niimg-like can go through `check_niimg` so can be an `img` or a string pointing to a nifti file plus the variation with the list of nifti filenames. The URL we added in a lot of docstrings is here: http://nilearn.github.io/building_blocks/manipulating_mr_images.html#niimg.

File: nilearn/_utils/niimg.py (hunk 23 above)
Comment (AlexandreAbraham) [inline nilearn/_utils/niimg.py]: OK. The official nibabel way to do that is `np.array(img.dataobj)`.

Comment (GaelVaroquaux) [inline nilearn/_utils/niimg.py]: > OK. The official nibabel way to do that is np.array(img.dataobj).

I don't think that it was possible when we wrote this code. I remember
having a lengthy discussion with Matthew Brett about this.

File: nilearn/_utils/niimg_conversions.py [hunk 24]
```diff
@@ -5,49 +5,14 @@
...
-    if hasattr(img, 'shape'):
-        shape = img.shape
-    else:
-        shape = img.get_data().shape
-    return shape
+from .niimg import (_safe_get_data, load_img, new_img_like,
+                    short_repr)
 
 
 def _check_same_fov(img1, img2):
```
Comment (banilo) [inline nilearn/_utils/niimg_conversions.py]: The same of this function is not exacgtly self-explanatory.

File: nilearn/_utils/testing.py [hunk 25]
```diff
@@ -17,14 +17,13 @@
...
 import scipy.linalg
-
-from nibabel import Nifti1Image
 import nibabel
 
 from .. import datasets
 from .. import masking
 from . import logger
 
+
```
Comment (banilo) [inline nilearn/_utils/testing.py]: Merge conflicts?

File: nilearn/image/resampling.py [hunk 26]
```diff
@@ -12,9 +12,10 @@
 import numpy as np
 import scipy
 from scipy import ndimage, linalg
-from nibabel import Nifti1Image
 
 from .. import _utils
+from .._utils import new_img_like
+
```
Comment (banilo) [inline nilearn/image/resampling.py]: Merge conflicts?

File: nilearn/_utils/niimg.py (hunk 23 above)
Comment (AlexandreAbraham) [inline nilearn/_utils/niimg.py]: OK. I have 2 versions of nibabel: 1.3.0 (has no cache) and 2.0 (has cache and dataobj) so I don't know which version is targeted by this code. I try to do the mean on 10 images from ADHD. I limited the memory. Here is the script:
```
from nilearn import datasets
from nilearn.image import mean_img
import resource


def limit_memory(maxsize):
    soft, hard = resource.getrlimit(resource.RLIMIT_AS)
    resource.setrlimit(resource.RLIMIT_AS, (maxsize, hard))


limit_memory(165152864 * 5)
data = datasets.fetch_adhd(n_subjects=10)
mean_img(data.func)
```

Given that paths are passed to `mean_img`, the caching system should not change anything between the versions. Both should be able to do the mean given the memory limit as they only need to load images one by one.

The result is that they both endup in memory error after 8 images. There must be a leak somewhere but I can't find out where. I'll keep the code as-is but we should maybe investigate that and add some tests.

Comment (GaelVaroquaux) [inline nilearn/_utils/niimg.py]: > OK. I have 2 versions of nibabel: 1.3.0 (has no cache) and 2.0 (has cache and
> dataobj) so I don't know which version is targeted by this code.

1.3.0 is targetted. 2.0 wasn't out when we wrote this code.

Comment (AlexandreAbraham) [inline nilearn/_utils/niimg.py]: Do you think that we should open an issue about potential memory leaks?

Comment (GaelVaroquaux) [inline nilearn/_utils/niimg.py]: > Do you think that we should open an issue about potential memory leaks?

Yes. With the example that you gave. It's a good one.

File: nilearn/_utils/niimg.py [hunk 27]
```diff
@@ -0,0 +1,157 @@
...
+
+    new_img: nibabel.SpatialImage
+        An image which has the same type as the reference image.
+    """
+    # XXX If the niimg is a list of 3D images, we don't need to load them all
+    # But this is a bit ugly
+    if hasattr(ref_img, "__iter__") and not isinstance(ref_img, string_types):
+        ref_img = next(ref_img)
+    ref_img = load_img(ref_img)
+    header = ref_img.get_header() if copy_header else None
```
Comment (GaelVaroquaux) [inline nilearn/_utils/niimg.py]: You should reinitialize a few entries of the header. For instance, the cal_max and cal_min. But you should actually go over the nifti documentation to check what the different fields of the header: not all should be copied.

File: nilearn/_utils/niimg.py (hunk 22 above)
Comment (AlexandreAbraham) [inline nilearn/_utils/niimg.py]: >  Is this added functionality actually used in the code?

Yes, but I can easily fix these cases. However, if one day we want to expose it to the user, I guess that eating niimgs will be mandatory.

File: nilearn/_utils/testing.py (hunk 25 above)
Comment (AlexandreAbraham) [inline nilearn/_utils/testing.py]: No, I just added a blank line for cosmetic pruprose.

File: nilearn/_utils/niimg.py (hunk 22 above)
Comment (GaelVaroquaux) [inline nilearn/_utils/niimg.py]: > Yes, but I can easily fix these cases. However, if one day we want to
> expose it to the user, I guess that eating niimgs will be mandatory.

The danger is that I have the impression here that you are exhausting the
first image of the iterator. So this code has a nasty side effect.

File: nilearn/_utils/niimg.py (hunk 27 above)
Comment (AlexandreAbraham) [inline nilearn/_utils/niimg.py]: I'd rather not do that in this PR. It has already gone far beyond what I wanted to do and I guess that the question of which header to keep / update / reset will be subject to debate. For the moment, I mimic the behavior of the current state of nilearn.

File: nilearn/_utils/niimg.py (hunk 22 above)
Comment (AlexandreAbraham) [inline nilearn/_utils/niimg.py]: In fact, this code works with a list but not with a generator. This will be the subject of another PR. In the current state, nilearn is absolutely not iterator/generator friendly as first elements are exhausted here and there. My goal with this PR was to clean the code and prepare the ground for further refactorings. I can deal with these problems in this PR of course but it will take more time.

Comment (AlexandreAbraham) [inline nilearn/_utils/niimg.py]: Some code like that may help but there will still be a side effect on the generator provided:
```
def _iter_first(first, iterable):
    yield first
    for i in iterable:
        yield i


def sneak_peek(iterable):
    first = next(iterable)
    return _iter_first(first, iterable), first
```

Comment (GaelVaroquaux) [inline nilearn/_utils/niimg.py]: > Some code like that may help but there will still be a side effect on the
> generator provided:

I think that this is better than the current situation, which is too
fragile.

File: nilearn/_utils/niimg.py (hunk 27 above)
Comment (GaelVaroquaux) [inline nilearn/_utils/niimg.py]: > I'd rather not do that in this PR. It has already gone far beyond what I wanted
> to do and I guess that the question of which header to keep / update / reset
> will be subject to debate. 

Well, the problem is that not resetting some values will give you bugs
that are very hard to debug with the softwares that respect these values.

I suggest that you put 'cal_min', 'cal_max' to zero (which means None in
Nifti terms), or better, to the min and max.

In general, I might even do the following:
{'scl_slope':0, 'scl_inter':0, 'glmax':0, 'glmin':0, 'cal_max':0,
'cal_min':0}

Once again: these things are really tricky, and if you don't do it right,
you'll end up with bugs. I am actually very weary of the nifti header.

File: nilearn/_utils/niimg.py (hunk 23 above)
Comment (lesteve) [inline nilearn/_utils/niimg.py]: > 1.3.0 is targetted.

then README.rst and nilearn/version.py should be updated. We are currently requiring nibabel 1.1.0 [there](https://github.com/nilearn/nilearn#dependencies).

Comment (GaelVaroquaux) [inline nilearn/_utils/niimg.py]: >     1.3.0 is targetted.

> then README.rst and nilearn/version.py should be updated. We are currently
> requiring nibabel 1.1.0 there.

OK, my answer was wrong. The question was: is it version 1.3.0 or version
2.0 that is targetted. I replied 1.3.0 but I should have replied 2.0.

Comment (AlexandreAbraham) [inline nilearn/_utils/niimg.py]: Oh we upgrade to 2.0, cool !

Comment (GaelVaroquaux) [inline nilearn/_utils/niimg.py]: Sorry, typo (doing too much email), I meant 1.1. Of course we don't update to 2.0. It's way too recent.

Comment (AlexandreAbraham) [inline nilearn/_utils/niimg.py]: I know ;).

**Comment by AlexandreAbraham** [comment]
Squashed last commits.

**Comment by AlexandreAbraham** [comment]
Sidenote: doc must be rewritten but I will do that after @GaelVaroquaux's #483 has been merged.

File: nilearn/_utils/niimg.py [hunk 28]
```diff
@@ -0,0 +1,144 @@
...
+    if data.dtype == bool:
+        data = as_ndarray(data, dtype=np.int8)
+    header = None
+    if copy_header:
+        header = copy.copy(ref_img.get_header())
+        header['scl_slope'] = 0.
+        header['scl_inter'] = 0.
+        header['glmax'] = 0.
+        header['cal_max'] = np.max(data) if data.size > 0 else 0.
+        header['cal_max'] = np.min(data) if data.size > 0 else 0.
```
Comment (GaelVaroquaux) [inline nilearn/_utils/niimg.py]: Could you use the nibabel update header function here (with a try/except if it's not available in old nibabel versions).

Comment (AlexandreAbraham) [inline nilearn/_utils/niimg.py]: So, apparently, update_header only takes care of shape/affines and is called automatically at image creation. If we want things to be done the right way, we should implement our own function to update these values.

File: nilearn/plotting/img_plotting.py (hunk 19 above)
Comment (lesteve) [inline nilearn/plotting/img_plotting.py]: Still don't get this.

File: nilearn/_utils/niimg.py (hunk 28 above)
Comment (GaelVaroquaux) [inline nilearn/_utils/niimg.py]: > So, apparently, update_header only takes care of shape/affines and is
> called automatically at image creation. If we want things to be done
> the right way, we should implement our own function to update these
> values.

OK, then let's keep what you have.

File: nilearn/plotting/img_plotting.py (hunk 19 above)
Comment (AlexandreAbraham) [inline nilearn/plotting/img_plotting.py]: What happens is that this template is added to the plots using `add_overlay`. However, `add_overlay` has an epsilon threshold used to remove numerical noise. This threshold is applied to the image, and thus a new image is created using new_img_like. However, the empty constructor in the template forbids allocation using this code:
`ref_img.__class__(data, affine, header=header)`. This is why I chose a polymorphic constructor to solve this problem.

Several options:
* The polymorphic constructor
* Have a custom type that does not allow modifications
* Use a clever code to fallback to Nifti1Image if a constructor fails in `new_img_like`
* Make the template derive from Nifti1Image and make clever code to find a suitable constructor in super classes in `new_img_like`
* Maybe other options I don't see.

File: nilearn/tests/test_niimg_conversions.py [hunk 29]
```diff
@@ -23,26 +23,30 @@
...
+        self.my_affine = np.ones((4, 4))
 
     def get_data(self):
         return self.data
 
     def get_affine(self):
-        return self.affine
+        return self.my_affine
+
+    @property
```
Comment (banilo) [inline nilearn/tests/test_niimg_conversions.py]: Why is that necessary?

Comment (lesteve) [inline nilearn/tests/test_niimg_conversions.py]: I guess so that you can use `bla.shape` rather than `bla.shape()`. Also IIRC properties are not settable by default so that trying `bla.shape = (3, 2, 1)` will fail.

Comment (AlexandreAbraham) [inline nilearn/tests/test_niimg_conversions.py]: :octocat: agrees with "le steve"

Comment (banilo) [inline nilearn/tests/test_niimg_conversions.py]: > agrees with "le steve"

hey...that's my trick
